jaxlib/pytree.cc

```diff
@@ -0,0 +1,616 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+// Caution: this code uses exceptions. The exceptions use is local to the
+// binding code and the idiomatic way to emit Python exceptions.
+
+#include <memory>
+#include <stdexcept>
+#include <utility>
+#include <vector>
+
+#include "absl/algorithm/container.h"
+#include "absl/container/flat_hash_map.h"
+#include "absl/hash/hash.h"
+#include "absl/memory/memory.h"
+#include "absl/strings/str_format.h"
+#include "absl/strings/str_join.h"
+#include "absl/synchronization/mutex.h"
+#include "include/pybind11/pybind11.h"
+#include "include/pybind11/pytypes.h"
+#include "include/pybind11/stl.h"
+
+
+namespace jax {
+
+namespace py = pybind11;
+
+// Registry of custom node types.
+class CustomNodeRegistry {
+ public:
+  struct Registration {
+    // The Python type object, used to identify the type.
+    py::object type;
+    // A function with signature: object -> (iterable, aux_data)
+    py::function to_iterable;
+    // A function with signature: (aux_data, iterable) -> object
+    py::function from_iterable;
+  };
+
+  // Registers a new custom type. Objects of `type` will be treated as container
+  // node types in PyTrees.
+  static void Register(py::object type, py::function to_iterable,
+                       py::function from_iterable);
+
+  // Finds the custom type registration for `type`. Returns nullptr if none
+  // exists.
+  static const Registration* Lookup(py::handle type);
+
+ private:
+  static CustomNodeRegistry* Singleton();
+
+  struct TypeHash {
+    size_t operator()(const py::object& t) const { return py::hash(t); }
+  };
+  struct TypeEq {
+    bool operator()(const py::object& a, const py::object& b) const {
+      return a.equal(b);
+    }
+  };
+  absl::Mutex mu_;
+  absl::flat_hash_map<py::object, std::unique_ptr<Registration>, TypeHash,
+                      TypeEq>
+      registrations_ GUARDED_BY(mu_);
+};
+
+/*static*/ CustomNodeRegistry* CustomNodeRegistry::Singleton() {
+  static auto* registry = new CustomNodeRegistry;
+  return registry;
+}
+
+/*static*/ void CustomNodeRegistry::Register(py::object type,
+                                             py::function to_iterable,
+                                             py::function from_iterable) {
+  CustomNodeRegistry* registry = Singleton();
+  absl::MutexLock lock(&registry->mu_);
+  auto registration = absl::make_unique<Registration>();
+  registration->type = type;
+  registration->to_iterable = std::move(to_iterable);
+  registration->from_iterable = std::move(from_iterable);
+  auto it = registry->registrations_.emplace(type, std::move(registration));
+  if (!it.second) {
+    throw std::logic_error("Duplicate custom PyTreeDef type registration.");
```

Comment:
nit: invalid_argument (since users can register their own types right?), and consider including 'type'